drivers/platform/stm32/gpio.c

```diff
@@ -0,0 +1,320 @@
+/******************************************************************************/
+
+/**
+ * @brief Prepare the GPIO decriptor.
+ * @param desc - The GPIO descriptor.
+ * @param param - The structure that contains the GPIO parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t _gpio_init(struct gpio_desc *desc,
+		   const struct gpio_init_param *param)
+{
+	int32_t ret = 0;
+	struct stm32_gpio_desc *extra = desc->extra;
+	struct stm32_gpio_desc *pextra = param->extra;
+	GPIO_InitTypeDef gis;
+
+	if (!param)
+		return -EINVAL;
+
+	/* enable gpio port in RCC */
+	if (pextra->port == GPIOA)
+		__HAL_RCC_GPIOA_CLK_ENABLE();
+#ifdef GPIOB
+	else if (pextra->port == GPIOB)
+		__HAL_RCC_GPIOB_CLK_ENABLE();
+#endif
+#ifdef GPIOC
+	else if (pextra->port == GPIOC)
+		__HAL_RCC_GPIOC_CLK_ENABLE();
+#endif
+#ifdef GPIOD
+	else if (pextra->port == GPIOD)
+		__HAL_RCC_GPIOD_CLK_ENABLE();
+#endif
+#ifdef GPIOE
+	else if (pextra->port == GPIOE)
+		__HAL_RCC_GPIOE_CLK_ENABLE();
+#endif
+#ifdef GPIOF
+	else if (pextra->port == GPIOF)
+		__HAL_RCC_GPIOF_CLK_ENABLE();
+#endif
+#ifdef GPIOG
+	else if (pextra->port == GPIOG)
+		__HAL_RCC_GPIOG_CLK_ENABLE();
+#endif
+#ifdef GPIOH
+	else if (pextra->port == GPIOH)
+		__HAL_RCC_GPIOH_CLK_ENABLE();
+#endif
+	else
+		return -EINVAL;
+
+	/* copy the settings to gpio descriptor */
+	desc->number = param->number;
+	extra->port = pextra->port;
+	extra->mode = pextra->mode;
+	extra->pull = pextra->pull;
+	extra->speed = pextra->speed;
+
+	/* configure gpio with user configuration */
+	gis.Pin = BIT(param->number);
+	gis.Mode = extra->mode;
+	gis.Pull = extra->pull;
+	gis.Speed = extra->speed;
+	HAL_GPIO_Init(extra->port, &gis);
+
+	return ret;
+}
+
+/**
+ * @brief Obtain the GPIO decriptor.
+ * @param desc - The GPIO descriptor.
+ * @param param - GPIO initialization parameters
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t stm32_gpio_get(struct gpio_desc **desc,
+		       const struct gpio_init_param *param)
+{
+	struct gpio_desc *descriptor;
+	struct stm32_gpio_desc *extra;
+	int32_t ret;
+
+	descriptor = (struct gpio_desc *)malloc(sizeof(*descriptor));
+	if (!descriptor) {
+		ret = -ENOMEM;
+		goto error;
+	}
+	extra = (struct stm32_gpio_desc*)malloc(sizeof(*extra));
+	if (!extra) {
+		ret = -ENOMEM;
+		goto error;
+	}
+
+	descriptor->extra = extra;
+	ret = _gpio_init(descriptor, param);
+	if(ret < 0)
+		goto error;
+
+	*desc = descriptor;
+
+	return 0;
+error:
+	free(extra);
+	free(descriptor);
+
+	return ret;
+}
+
+/**
+ * @brief Get the value of an optional GPIO.
+ * @param desc - The GPIO descriptor.
+ * @param param - GPIO Initialization parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t stm32_gpio_get_optional(struct gpio_desc **desc,
+				const struct gpio_init_param *param)
+{
+	if(param == NULL) {
+		*desc = NULL;
+		return SUCCESS;
+	}
+
+	return stm32_gpio_get(desc, param);
```

Comment:
Why not implementing this in gpio.c (the generic one)

Reply:
You mean why do we implement platform specific gpio_get_optional functions ? good question, historical reasons i guess